### PR TITLE
fix(core): Listing Cloud (S3, GCS, etc.) state IDs no longer crashes if there is a file at the root of the bucket, and files in GCS buckets are only listed within the prefix specified in the `state_backend_uri`

### LINKED ISSUE #8425: bug: `meltano state list` fails for the Google Cloud Storage backend if there's files at the root of the bucket
### Meltano Version

3.3.1

### Python Version

NA

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

NA

### Description

For example, if the contents of a bucket look like this

```
my-bucket
├── my_file.txt
└── state
    ├── state-id-1
    │   └── state.json
    └── state-id-2
        └── state.json
```

then `meltano state list` with `state_backend.uri` set to `gs://my-bucket/state` will fail with a `ValueError: not enough values to unpack` in

https://github.com/meltano/meltano/blob/d6997eb1e692978a7da56bffbf7d66d677027864/src/meltano/core/state_store/filesystem.py#L585

because bucket contents are iterated at the root of the bucket, without considering the configured prefix:

https://github.com/meltano/meltano/blob/d6997eb1e692978a7da56bffbf7d66d677027864/src/meltano/core/state_store/google.py#L129

---

The fix probably involves two things:

1. Skipping the blob filename if it doesn't contain a `/`
2. Passing `bucket=...` to `list_blobs`

We might also want to ensure the prefix doesn't have a leading slash.

### Code

_No response_

### LINKED ISSUE #8647: bug: `meltano state list` lists all files in S3 bucket, ignoring the prefix specified in `state_backend_uri`
### Meltano Version

3.4.2

### Python Version

3.12

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

NA

### Meltano Version

3.4.2

### Python Version

NA

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

NA

### Description

<details><summary>Outdated crash with files at the root of the bucket</summary>
<p>

then `meltano state list` with `state_backend.uri` set to `s3://my-bucket/state` will fail with a `ValueError: not enough values to unpack` in

https://github.com/meltano/meltano/blob/d6997eb1e692978a7da56bffbf7d66d677027864/src/meltano/core/state_store/filesystem.py#L585

</p>
</details> 

For example, if the contents of a bucket look like this

```
my-bucket
├── my_file.txt
└── state
    ├── state-id-1
    │   └── state.json
    └── state-id-2
        └── state.json
```

and the state backend URI setting is `s3://my-bucket/state`, the bucket contents are iterated at the root of the bucket, without considering the configured prefix:

https://github.com/meltano/meltano/blob/72454b4a973e279c383796bd37d5f8a6fd128f13/src/meltano/core/state_store/s3.py#L143-L146

So Meltano incorrectly tries to parse `my_file.txt` for state.

---

The fix probably two things passing passing `Prefix=` to `list_objects_v2`.

We might also want to ensure the prefix doesn't have a leading slash.

### Code

_No response_

### Link to Slack/Linen

https://meltano.slack.com/archives/C069CQNHDNF/p1721603396866689

### Related issues

- GCS issue: https://github.com/meltano/meltano/issues/8425

---

UPDATE (2024-07-22): The first problem described here (files at the bucket root make listing state IDs crash) was fixed by https://github.com/meltano/meltano/pull/8648.

### PR DESCRIPTION
**Pull request by jx2lee**
## Description

when any files were in root GCS bucket, `meltano state list` cmd failed.
details on [#8425](https://github.com/meltano/meltano/issues/8425)

## Related Issues

* Closes #8425

### CODE REVIEW

**Comment by jx2lee** [comment]
Could i need to additional unit test?

File: src/meltano/core/state_store/google.py [hunk 1]
```diff
@@ -127,7 +127,9 @@ def list_all_files(self) -> Iterator[str]:
         Yields:
             The next file in the backend.
         """
-        for blob in self.client.list_blobs(bucket_or_name=self.bucket):  # noqa: WPS526
+        for blob in self.client.list_blobs(
+            bucket_or_name=self.bucket, prefix=self.state_dir
+        ):  # noqa: WPS526
```
Comment (edgarrmondragon) [inline src/meltano/core/state_store/google.py:132]:
```suggestion
        for blob in self.client.list_blobs(  # noqa: WPS526
            bucket_or_name=self.bucket, prefix=self.state_dir
        ):
```

File: tests/meltano/core/state_store/test_filesystem.py [hunk 2]
```diff
@@ -616,4 +637,22 @@ def test_get_state_ids(self, subject: GCSStateStoreManager):
             for i in range(10)
         )
         assert set(subject.get_state_ids()) == {f"state_id_{i}" for i in range(10)}
-        subject.client.list_blobs.assert_called_once_with(bucket_or_name="meltano")
+        subject.client.list_blobs.assert_called_once_with(
+            bucket_or_name="meltano", prefix="state"
+        )
+
+    @pytest.mark.usefixtures("mock_client")
+    def test_get_state_ids_when_any_files_was_located__in_root(
```
Comment (edgarrmondragon) [inline tests/meltano/core/state_store/test_filesystem.py:645]:
```suggestion
    def test_get_state_ids_when_any_files_was_located_in_root(
```

**Comment by edgarrmondragon** [comment]
> Could i need to additional unit test?

This seems reasonably covered. Do you think there's a missing unit test?

**Comment by jx2lee** [comment]
@edgarrmondragon 
No, I was just wondering if there was anything else you want to add to code I wrote!

**Review by edgarrmondragon — APPROVED** [review]
LGTM!

Thank you contributing @jx2lee!

**Comment by edgarrmondragon** [comment]
https://github.com/all-contributors please add @jx2lee for reporting a submitting a pull request to fix this issue.

**Comment by edgarrmondragon** [comment]
@all-contributors please add @jx2lee for reporting a submitting a pull request to fix this issue.

**Comment by edgarrmondragon** [comment]
@all-contributors please add @jx2lee for reporting a submitting code to fix this issue.

**Pull request merged**